src/main/scala/tilelink/SRAM.scala

```diff
@@ -74,48 +75,100 @@ class TLRAM(
         LogicalModuleTree.add(parentLTN, sramLogicalTreeNode)
     }
 
-    /* This block uses a two-stage pipeline; A=>D
-     * Both stages vie for access to the single SRAM port.
-     * Stage D has absolute priority over stage A.
-     *   - read-modify-writeback for sub-lane access happens here
-     *   - writeback of correctable data happens here
-     *   - both actions may occur concurrently
-     * Stage A has lower priority and will stall if blocked
-     *   - read operations happen here
-     *   - full-lane write operations happen here
+    /* This block has a three-stage pipeline
+     * Stage A is the combinational request from TileLink A channel
+     * Stage R corresponds to an accepted request
+     * Stage D registers the result of an SRAM read (if any)
+     *
+     * The TileLink D channel response comes from
+     *   - stage D for corected reads or AMOs
+     *   - stage R for everything else
+     *   - However, to increase maximum operating frequency, the
+     *     stage R responses can be configured to come from stage D
+     *
+     * For sub-ECC granule writes and atomic operations:
+     *   - stage A sets up the read for the old data value
+     *   - stage R is used to gather the result from SRAM to registers
+     *   - stage D corrects ECC, applies the ALU, and sets up SRAM write
+     *
+     * For super-ECC granule writes:
+     *   - stage A sets up the write
+     *
+     * For reads:
+     *   - stage A sets up the read
+     *   - stage R drives the uncorrected data with valid based on ECC validity
+     *   - stage D sets up the correction, if any
+     *
+     * When stage D needs to perform a write (AMO, sub-ECC write, or ECC correction):
+     *   - there is a WaW or WaR hazard vs. the operation in stage R
+     *     - for sub-ECC writes and atomics, we ensure stage R has a bubble
+     *     - for ECC correction, we cause stage R to be replayed (and reject stage A twice)
```

Comment:
What's the rationale for not using replay for both cases?  If using a bubble instead of a replay would improve performance, I'd understand.  But since there's a structural hazard from the D stage on the following cycle anyway, is the bubble actually better than the replay?

Reply:
You're right that the performance in terms of latency and throughput would be the same. However, if I replay it requires reading the SRAM twice. I assumed that sub-ecc-granule writes might be common, so paying 2x power for them was unwise.